omnigibson/utils/bddl_utils.py

```diff
@@ -89,9 +127,810 @@ def process_single_condition(condition):
+            entity (None or DatasetObject or BaseSystem): If specified, the BDDL entity to wrap. If not
+                specified, will initially wrap nothing, but may dynamically reference an actual object or system
+                if it exists in the future
+        """
+        # Store synset and other info, and pass entity internally
+        self.object_scope = object_scope
+        self.synset = "_".join(self.object_scope.split("_")[:-1])
+        self.is_system = self.synset in SUBSTANCE_SYNSET_MAPPING
+
+        # Infer the correct category to assign, special casing agents
+        if entity is not None and isinstance(entity, BaseRobot):
+            self.og_categories = ["agent"]
+        else:
+            self.og_categories = [SUBSTANCE_SYNSET_MAPPING[self.synset]] if self.is_system else \
+                OBJECT_TAXONOMY.get_subtree_igibson_categories(self.synset)
+
+        super().__init__(obj=entity)
+
+    def _find_valid_object(self):
+        """
+        Internal helper function to find the first valid simulator object whose category is one of @self.og_categories,
+        and has not been mapped to a BDDLEntity yet
+
+        Returns:
+            None or DatasetObject: If found, the valid object matching a category from @self.og_categories and not
+                mapped
+        """
+        for category in self.og_categories:
+            for obj in og.sim.scene.object_registry("category", category, default_val=[]):
+                if isinstance(obj, DatasetObject) and obj.bddl_object_scope is None:
+                    # Found valid one, return it
+                    return obj
+
+    def exists(self):
+        """
+        Checks whether the entity referenced by @synset exists. Note: this dynamically mutates self.wrapped_obj, and
+        potentially removes it or adds a reference if the entity no longer / now exists.
+
+        Returns:
+            bool: Whether the entity referenced by @synset exists
+        """
+        if self.wrapped_obj is None:
+            # If system, check to see if active or not and grab it if so
+            if self.is_system:
+                if is_system_active(self.og_categories[0]):
+                    self.wrapped_obj = get_system(self.og_categories[0])
+            # Otherwise, is object, check to see if any valid one exists and grab it if so
+            else:
+                found_obj = self._find_valid_object()
+                if found_obj is not None:
+                    found_obj.bddl_object_scope = self.object_scope
+                    self.wrapped_obj = found_obj
+        else:
+            # Check to see if entity no longer exists
+            if self.is_system:
+                if not is_system_active(self.og_categories[0]):
+                    self.wrapped_obj = None
+            # Otherwise, is object, check to see if there are no valid ones
+            else:
+                if og.sim.scene.object_registry("name", self.wrapped_obj.name) is None:
```

Comment:
nit: maybe this should be a separate utility function? I am not sure.

Reply:
outdated